### DISCUSSION TITLE
Security details + best practices

### DISCUSSION THREAD

**Discussion by jodyheavener** [microsoft/vscode-discussions#8] [Extension Development QnA] [answered]
I'm working on an extension that deals a lot with sensitive information (i.e. secrets) and I've been tasked to try and understand some of the details around security in an extension. Particularly I'm curious about how/if data is stored, and when.

**Secrets**

For storing sensitive data, such as API keys, I understand that WorkspaceConfiguration and ExtensionContext states are not suitable, but the Secrets API and [SecretStorage](https://code.visualstudio.com/api/references/vscode-api#SecretStorage) is okay for this. I am unable to find any documentation or implementation details beyond this announcement paragraph where it states

> This allows extensions to store sensitive information in the OS credential manager or keystore.

How secure is this? If another process knows where to look, could it find this data? Furthermore (and maybe this is more appropriate for a feature request Issue) has VS Code considered allowing outside vendors to sync and store data? I know 1Password would be interested in providing a "secure storage" method.

Any implementation details or definitive information is greatly appreciated.

**Commands**

When it comes to commands, I understand that all registered commands (user facing and internal) are global, and any extension that knows the right ID could execute a command. Are there any mechanisms or practices that could be used to prevent a malicious extension from repeatedly executing a command?

I could see an internal command registering under a random string each time the extension starts to limit outside extensions from executing it.

**Logging**

My extension uses an [OutputChannel](https://code.visualstudio.com/api/references/vscode-api#OutputChannel) to log debugger details. What is the exposure of these logs? Are they stored to a file on disk, or are they limited only to VS Code in memory? I see there's [an issue](https://github.com/microsoft/vscode/issues/65108) about retrieving from an existing channel, which, if implemented would make it pretty easy to fetch another extension's logs. 

Finally, when using `showInputBox` or `showQuickPick`, does this received input get logged anywhere, or is it passed directly back to the extension?

---

I realize there are a bunch of fairly specific questions here, but if anyone knows any definitive answers here I think this should serve as a good discussion for future folks wondering as well! Additionally, if there is anything security-wise that I have not asked and should be taken into account, I'd love to hear it. Thanks!

**Comment by isidorn** [comment, marked as answer]
Thanks for creating this great discussion. Let me try to answer your questions:

Commands - there is no mechanism to prevent a malicious extension from executing your commands. In fact malicious extensions running under VS Code could do much worse - change the source code of your extension by using fs api, so any prevention you do here would be futile. In the future we plan to investigate more into sandboxing extensions, and then discussing this more would make good sense.

Logging - the output is stored in a file as an implementation detail. We do not expose the file to extensions via API. For example use command `Developer: Open Extensions Log Folder` there you will see output files. Yes, that issue is open but currently there are no plans to fix this. However same as above, if there is a malicious extension it might find some smart evil way to get the content of your logs - eg fs api to read the content of those files. 

`showInputBox`, `showQuickPick` - input is not logged anywhere. Returned directly to extension.

Secrets - @tylerleonhardt can you help here as the author of our SecretStorage api. Thanks!

**Reply by TylerLeonhardt** [reply]
### Secret Storage

Secrets are stored in the keyring of the OS you're on.

We use http://github.com/atom/node-keytar

> A native Node module to get, add, replace, and delete passwords in system's keychain. On macOS the passwords are managed by the Keychain, on Linux they are managed by the Secret Service API/libsecret, and on Windows they are managed by Credential Vault.

Additionally we encrypt the strings before adding them to the keyring to make it harder for other processes and extensions that also have access to the keyring to read the secrets stored with the SecretStorage API.

**Reply by jodyheavener** [reply]
Fantastic - thanks both of you for taking the time to respond to these questions!

**Comment by ronakj** [comment]
@TylerLeonhardt A couple of questions for secret storage API -
- Are the secrets stored locally when using the remote extensions?
- When working with Codespaces in browser or when using VSCode server, will the secrets also be stored in browser? Or will they be stored in remote VM/Container?

**Comment by TylerLeonhardt** [comment]
@ronakj Here's the break down:

* using VS Code desktop with local or any remote workspace: secrets are stored on the client side
* Codespaces, github.dev, or vscode.dev: secrets are stored in browser local storage encrypted

**Reply by ronakj** [reply]
Thanks!
